File: CPP/ByRound/1766/1766_C_Hamiltonian_Wall.cpp
// Problem: CF 1766 C - Hamiltonian Wall
// https://codeforces.com/contest/1766/problem/C

/*
Purpose: Determine if a Hamiltonian path exists on a 2xM grid such that all black cells are visited exactly once, 
         and no white cells are visited. The path must follow adjacent cells (sharing a side) and must cover all black cells.

Algorithms/Techniques: 
- Recursive backtracking with memoization to explore valid paths
- Try starting from either row (0 or 1) and column 0
- At each step, decide whether to move right or potentially change rows (if the current column has no black cell in the other row)

Time Complexity: O(M), where M is the number of columns. Each cell is visited at most once due to the recursive structure.
Space Complexity: O(M), due to recursion stack depth in worst-case scenarios (linear recursion).
*/

#include <iostream>
#include <string>

using namespace std;
int n;
string s[2];

// Function to check if a valid Hamiltonian path exists starting from (i, j)
// Returns 1 if path exists, 0 otherwise
int g(int i, int j) {
  // Base case: reached end of column
  if (j == n)
    return 1;
  // If current cell is white, it's invalid
  if (s[i][j] != 'B')
    return 0;
  // If the other row at this column is also black, we must switch rows
  if (s[1 - i][j] == 'B')
    return g(1 - i, j + 1);
  // Otherwise, we continue in the same row
  return g(i, j + 1);
}

int main() {
  // Read number of columns
  for (cin >> n; cin >> n >> s[0] >> s[1];)
    // Try starting from both rows (0 and 1) and output YES if any path works
    cout << (g(0, 0) | g(1, 0) ? "YES" : "NO") << endl;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/C/1766_C_Hamiltonian_Wall.py
# Problem: CF 1766 C - Hamiltonian Wall
# https://codeforces.com/contest/1766/problem/C

"""
Code Purpose:
This code determines whether a Hamiltonian path exists on a 2-row, m-column grid such that:
1. The path visits every black cell exactly once.
2. The path only moves through adjacent cells (sharing a side).
3. No white cells are visited.
4. Each column has at least one black cell.

The algorithm analyzes the pattern of black and white cells in each column:
- If both rows in a column are the same (both B or both W), the column is processed differently.
- If the rows in a column differ, the algorithm tracks a 'state' variable to determine how to proceed.
- The solution checks that there are at most 2 distinct 'types' of columns in the path (represented by the set s).

Algorithms/Techniques:
- Greedy approach with state tracking
- Column-by-column analysis
- Set-based tracking of column types

Time Complexity: O(m) where m is the number of columns
Space Complexity: O(1) as only a constant amount of extra space is used (set s and variables i, x)
"""

I = input
for _ in [0] * int(I()):
    I()  # Read m (number of columns), but don't use it
    s = {i := 0}  # Initialize a set s with 0, and i = 0
    for x in zip(I(), I()):  # Iterate over pairs of characters from two rows
        if x[0] == x[1]:  # If both cells in the column are the same
            i ^= 1  # Toggle the value of i (0 -> 1 or 1 -> 0)
        else:  # If cells in the column differ
            s |= {x[i]}  # Add the character from the row determined by i to set s
    print("YNEOS"[len(s) > 2 :: 2])  # Print YES if len(s) <= 2, otherwise NO


# https://github.com/VaHiX/CodeForces/